### ISSUE TITLE
Feature request: Add an option to allow methods and (arrow function) properties to be mixed (sort-comp)

### ISSUE THREAD

**Issue by mckravchyk** [jsx-eslint/eslint-plugin-react#3439]
It seems like the plugin requires "everything-else" methods to be placed after "everything-else" arrow function properties. I understand that this is probably a basic rule that requires properties to be defined first. However, arrow functions are most commonly used as event handlers. Personally it makes more sense to me when the event handlers are closer to the render method. More importantly - it's sometimes the case that an arrow function must be used for something other than a DOM event handler and placing it first may not necessarily be the best position choice.

**Comment by mckravchyk** [comment]
Checking for arrow functions specifically is not a requirement for me (I understand this may be hard or not possible without TS Eslint) - it would just suffice if properties and methods could be mixed.

**Comment by ljharb** [comment]
Arrow functions should never go in class fields anyways - instead, you should do this:
```js
onEvent = this.onEvent.bind(this);
onEvent() {}
```
to maximize reusability, sharing of functions, and mockability in tests.

**Comment by mckravchyk** [comment]
I don't agree. In your example the method is still bound to the instance, just like an arrow function is, thus - not fully re-usable. It is theoretically, but in practice something like this is confusing when using it externally and also a foot gun internally that TypeScript would not shield against (what if you forget to bind?). Perhaps using a static method like `Comp.onEvent(instance, e) ` used like  `onEvent={(e) => Comp.onEvent(this, e)}` would solve this problem, but then why not make it a regular function and why should it not be in a different file anyway - which is exactly what I would do if I wanted to have a re-usable event handler.

**Comment by ljharb** [comment]
The method itself is the one on the prototype - which is reusable - it's only the shallow bind-proxy that's in the field.

TS is irrelevant; if you forget to bind the method wouldn't work. You could (and should) also make it a *private* method, which would mean that if you forget to bind it, it throws.

I don't mean reusable across components - i mean reusable across component *instances*. Functions are optimized when they're invoked multiple times - the same component might have dozens of instances, and that gets you an optimized function much faster than one that's recreated from scratch every `new`.

**Comment by mckravchyk** [comment]
That's a good point, though is it that much of a performance overhead to let them re-initialize? I know that globally it definitely adds up to some otherwise unnecessary computations, but it may be still low compared to resources used by the app in general.

If you forget to bind it wouldn't work, but that would not pop up until the runtime. I meant that it's something that cannot be statically checked, unless there's some Eslint plugin for this.

Given the lack of static type checking I still prefer arrow functions for such situations, I think the micro-performance overhead is worth it in this situation and I'm saying that as someone that's all in on performance. I will consider the approach with static methods I mentioned in my previous comment - then TS would not allow me to use `this` as the instance type and force to use the instance passed as the argument.

**Comment by ljharb** [comment]
That's fine, that's what unit tests are for.

It's not just performance of course; it's conceptually incorrect to have things that aren't data stored in per-instance fields.

**Comment by mckravchyk** [comment]
> That's fine, that's what unit tests are for.

Not everyone can afford those on every component :) Especially when prototyping something, specs change too often and you  still want the prototype to be as bug-free as possible with minimal time investment.

Yes, I agree about it being conceptually incorrect. I will probably start using static methods and pass the instance as the argument from now on. Thanks for your feedback.

**Comment by ljharb** [comment]
What everyone can’t afford is any untested component, ever :-) either way, if you’re content without tests then I’m not sure why the linter is needed?

**Comment by ljharb** [comment]
While I disagree very very strongly, if having tests is a waste of time then surely running the linter is a waste of time.

**Comment by mckravchyk** [comment]
Maybe I should not say: "waste of time" - it's just that for me, the time is better spent elsewhere. Of course, I'm saying this working with a statically typed language. If TypeScript or others like it did not exist, perhaps I would have a different opinion about the topic. I have done some small fixes once on a codebase developed by someone else, JS, no tests at all  - every time I made a change I had to stress about something breaking. I do not experience that anxiety when I use TS.

I don't think that complying with the linter is anywhere on the same scale as 100% test coverage. It's a small time investment and one of the benefits is that it automates away the need to think about code style, so it actually saves more time than it takes. I do invest a significant amount of time to ensure that the project is properly typed though but I don't see it as being a contradiction to not unit test every single thing,  quite the opposite - using the type system allows me to worry less about covering everything. I don't say that I don't unit test, but we talk about components, and again, to me the components are usually dead simple and TypeScript will cover most of the things that would otherwise be prone to break - in most components there's no complex logic (at least that's the way I do it).
